### PR TITLE
Security audit and production readiness fixes for Tansu smart contrac…

### PR DESCRIPTION
**Pull request by Septimus4**
Strengthens the governance contract by addressing unsafe patterns, clarifying rules, and adding protective checks. It improves security, maintainability, and auditability.

I focused only on code inside contracts/tansu/src
For Point 7 specifically, my understanding:
Total votes used as denominator included abstentions. That means abstaining acted like a silent “no.”
Example: with 10 voters, 4 yes, 0 no, 6 abstain → 40% yes → proposal fails.

Doc description (before my update): majority was described as ratio of yes over (yes + no), meaning abstentions are ignored. Same case → 4 yes, 0 no → 100% yes → proposal passes.

The intent is stricter governance (requirement that a majority of all eligible voters must approve), then code was correct and docs needed updating. If the intent is looser governance (only count explicit yes/no votes), then docs were correct and code needed changing. @tupui, I went with the code a source of truth and updated the doc, I can do the other way around.

1. **Maintainer authorization on `execute`**
   Added explicit checks so only designated maintainers can trigger contract upgrades.
   Why: Prevents unauthorized execution that could bypass governance decisions.

2. **Voting deadline enforcement**
   Enforced timestamp comparison against `env.ledger().timestamp()`.
   Why: Eliminates ambiguity about late votes and ensures proposals close predictably.

3. **Anonymous vote and seed validation**
   Enforced strict equality between votes and seeds arrays, with explicit errors.
   Why: Protects against index errors, malformed submissions, and tally corruption.

4. **Removal of unsafe `unwrap()`**
   Replaced all reachable `unwrap()` calls with guarded branches and error returns.
   Why: Prevents contract panics that could be exploited for denial of service.

5. **Logical operator corrections**
   Replaced bitwise `|` and `&` with logical `||` and `&&`.
   Why: Ensures correct boolean evaluation and short-circuiting, reducing logic errors.

6. **Governance threshold validation**
   Rejected thresholds of 0 or greater than the number of maintainers.
   Why: Prevents governance lockups or trivial approvals that undermine the system.

7. **Majority formula clarification**
   Documented and aligned behavior on how abstentions affect vote outcomes.
   Why: Removes ambiguity between documentation and implementation, ensuring predictable governance.

8. **Timestamp boundary tests**
   Added explicit test for voting at exactly the deadline tick.
   Why: Clarifies off-by-one edge cases and prevents disputes around deadline interpretation.

9. **Pattern simplification in `vote()`**
   Rewrote nested `if let` constructs into clear conditionals.
   Why: Improves readability, reduces audit complexity, and ensures stability across Rust versions.

10. **BLS commitment verification**
    Added tests with known vectors and documented cryptographic assumptions (curve, RNG, serialization).
    Why: Ensures anonymity mechanisms are verifiable and built on explicit, auditable assumptions.

11. **Pagination and gas limits**
    Introduced hard caps on page sizes and added gas-scaling tests.
    Why: Protects against resource exhaustion and DoS vectors.

### CODE REVIEW

**Comment by tupui** [comment]
Hi @Septimus4 thanks for doing that!

Can you rebase on #240 or have a look at it? I added a bunch of things.

**Comment by Septimus4** [comment]
@tupui, rebased and I added a commit to improve the pause logic:
enforce pause checks on all admin state changes (only execute upgrades when unpaused),
updated constructor to save one storage read/write op,
include wasm_hash in the collateral update event for parity

File: contracts/tansu/src/contract_tansu.rs [hunk 1]
```diff
@@ -9,9 +9,8 @@ impl TansuTrait for Tansu {
     /// * `env` - The environment object
     /// * `admin` - The admin address
     fn __constructor(env: Env, admin: Address) {
```
Comment (tupui) [inline contracts/tansu/src/contract_tansu.rs:11]: Ok so actually what we need to do here is to initialise as paused by just calling `Self::pause(env, admin, true);` or manually might be just better/simpler
```rust
env.storage()
            .instance()
            .set(&types::DataKey::Paused, &true);
```
Because in the end this is the init and we do not need the event. It's more or less useless.

The require_auth for the admin is sort of useless but does not hurt to have it there I suppose. Because when you deploy that runs atomically and you need to sign. So it's a given that you are authenticated.

File: contracts/tansu/src/contract_tansu.rs [hunk 2]
```diff
@@ -91,6 +97,8 @@ impl TansuTrait for Tansu {
     /// * `domain_contract` - The new domain contract
     fn set_domain_contract(env: Env, admin: Address, domain_contract: types::Contract) {
         auth_admin(&env, &admin);
+        // Enforce pause semantics for state-changing admin ops.
+        Self::require_not_paused(env.clone());
```
Comment (tupui) [inline contracts/tansu/src/contract_tansu.rs:101]: No on the contrary, if there is an issue with these contracts we want to be able to do such admin ops in a paused state. The admin would pause the contract to prevent anything bad from happening then be able to adjust these.

File: contracts/tansu/src/contract_tansu.rs [hunk 3]
```diff
@@ -115,6 +123,8 @@ impl TansuTrait for Tansu {
     /// * `collateral_contract` - The new collateral contract
     fn set_collateral_contract(env: Env, admin: Address, collateral_contract: types::Contract) {
         auth_admin(&env, &admin);
+        // Enforce pause semantics for state-changing admin ops.
+        Self::require_not_paused(env.clone());
```
Comment (tupui) [inline contracts/tansu/src/contract_tansu.rs:127]: Same here

File: contracts/tansu/src/contract_tansu.rs [hunk 4]
```diff
@@ -150,6 +161,8 @@ impl TansuTrait for Tansu {
         new_admins_config: Option<types::AdminsConfig>,
     ) {
         auth_admin(&env, &admin);
+        // Proposing an upgrade mutates state; require not paused.
+        Self::require_not_paused(env.clone());
```
Comment (tupui) [inline contracts/tansu/src/contract_tansu.rs:165]: Even more important here. Admins would even pause the contract to do a migration to prevent anything bad from happening.

File: contracts/tansu/src/contract_tansu.rs [hunk 5]
```diff
@@ -195,6 +213,8 @@ impl TansuTrait for Tansu {
     /// * If there is no upgrade to approve
     fn approve_upgrade(env: Env, admin: Address) {
         let admins_config = auth_admin(&env, &admin);
+        // Approving an upgrade mutates state; require not paused.
+        Self::require_not_paused(env.clone());
```
Comment (tupui) [inline contracts/tansu/src/contract_tansu.rs:217]: Same no

File: contracts/tansu/src/contract_tansu.rs [hunk 6]
```diff
@@ -255,6 +275,8 @@ impl TansuTrait for Tansu {
             });
 
         if accept {
+            // Executing an upgrade should respect pause state.
+            Self::require_not_paused(env.clone());
```
Comment (tupui) [inline contracts/tansu/src/contract_tansu.rs:279]: Same no

File: website/docs/developers/governance.mdx [hunk 7]
```diff
@@ -56,14 +64,40 @@ The helper `build_commitments_from_votes` can be called in simulation to reprodu
...
+- **Protection against hasty decisions** - changes require overwhelming support
+
 Custom quorum logic (percentage, thresholds…) can be built off-chain before deciding whether to call `execute`.
 
+### Voting Eligibility
+
+To cast a vote on a proposal, users must:
+
+- Be registered as a member using `add_member`
+- Have one or more badges assigned for the project (see [Membership & Badges](./membership.mdx))
```
Comment (tupui) [inline website/docs/developers/governance.mdx:95]: People can have no badges, they just get weight as 1. Maybe I should revisit the values. I will think about that for some follow ups.

File: contracts/tansu/src/tests/test_security.rs
Comment (tupui) [inline contracts/tansu/src/tests/test_security.rs:1]: These belong to the existing test files I think.

**Comment by Septimus4** [comment]
With all the rebase I had issues, I will open it updated and cleaned soon

**Comment by tupui** [comment]
Ok no worries 👍 thanks for following up.

My objective is to attempt mainnet this week. I sort of prepared the rest on Netlify, Cloudflare and the registrar. Only need to deploy the contract basically.
